src/libultra/libc/bcopy.s

```diff
@@ -0,0 +1,211 @@
+#include "PR/asm.h"
+#include "PR/regdef.h"
+
+.text
+
+LEAF(bcopy)
+    move    a3, a1
+    beqz    a2, ret
+    beq     a0, a1, ret
+    blt     a1, a0, goforwards
+    add     v0, a0, a2
+    bge     a1, v0, goforwards
+    b       gobackwards
+
+goforwards:
+    blt     a2, 0x10, forwards_bytecopy
+    and     v0, a0, 3
+    and     v1, a1, 3
+    beq     v0, v1, forwalignable
+
+forwards_bytecopy:
+    beqz    a2, ret
+    addu    v1, a0, a2
+99:
+    lb      v0, (a0)
+    addu    a0, a0, 1
+    sb      v0, (a1)
+    addu    a1, a1, 1
+    bne     a0, v1, 99b
+ret:
+    move    v0, a3
+    jr      ra
+
+forwalignable:
+    beqz    v0, forwards_32
+    beq     v0, 1, forw_copy3
+    beq     v0, 2, forw_copy2
+
+    lb      v0, (a0)
+    addu    a0, a0, 1
+    sb      v0, (a1)
+    addu    a1, a1, 1
+    addu    a2, a2, -1
+    b       forwards_32
+
+forw_copy2:
+    lh      v0, (a0)
+    addu    a0, a0, 2
+    sh      v0, (a1)
+    addu    a1, a1, 2
+    addu    a2, a2, -2
+    b       forwards_32
+
+forw_copy3:
+    lb      v0, (a0)
+    lh      v1, 1(a0)
+    addiu   a0, a0, 3
+    sb      v0, (a1)
+    sh      v1, 1(a1)
+    addiu   a1, a1, 3
+    addiu   a2, a2, -3
```

Comment:
Same as #1816
Both `addu a0, a0, 2` and `addiu a0, a0, 3` are essentially the same (besides the 2 vs 3 ofc).